framework/src/base/Assembly.C

```diff
@@ -3334,12 +3334,11 @@ Assembly::cacheResidual()
 {
   const std::vector<MooseVariableFEBase *> & vars = _sys.getVariables(_tid);
   for (const auto & var : vars)
-    for (MooseIndex(_residual_vector_tags) tag_index = 0; tag_index < _residual_vector_tags.size();
-         tag_index++)
-      if (_sys.hasVector(_residual_vector_tags[tag_index]._id))
-        cacheResidualBlock(_cached_residual_values[tag_index],
-                           _cached_residual_rows[tag_index],
-                           _sub_Re[tag_index][var->number()],
+    for (const auto & vector_tag : _residual_vector_tags)
+      if (_sys.hasVector(vector_tag._id))
+        cacheResidualBlock(_cached_residual_values[vector_tag._type_id],
```

Comment:
so `_residual_vector_tags` is not ordered based on `_type_id`?

Reply:
`_residual_vector_tags` is ordered by default. If I did not make this change, the code would still run fine.

But I do not want to make an unnecessary assumption. And so I made code more general